text/0000-cargo-workspace-deduplicate.md

```diff
@@ -0,0 +1,633 @@
+you're depending on exactly that copy on the local filesystem. This also means
+that if you depend on any workspace member via a `git` dependency later on it'll
+correctly pull in the other workspace members from the git repo. (note that some
+projects use `[patch]` to only write down `other-workspace-member = "0.2.3"` but
+this causes issues when crates later use git dependencies)
+
+If you never publish to crates.io, `path` is all you need. If crates eventually
+get published, though, they also need a `version` directive to know what version
+from crates.io you'll be depending on after the publication.
+
+Naturally, with a highly-interconnected workspace which may be relatively large,
+this leads to a lot of duplication very quickly. This is a lot of `path` and
+`version` directives that you've got to manage.
+
+## Duplication in crate versions
+
+A frequent pattern in Cargo workspaces which publish to crates.io is to have all
+the crate at the same semver version. These crates all move in lockstep during
+publication and get bumped at the same time.
+
+While a minor papercut this basically means that anyone and everyone who has a
+workspace of a lot of crates makes their own homebrew script for updating
+versions and managing updates/publications. It'd be quite convenient if we could
+standardize across the Rust ecosystem how to manage this information!
+
+## Duplication in crate metadata
+
+The last primary area of duplication that this proposal attempts to tackle is in
+crate metadata in the `[package]` section. This includes items such as:
+
+```toml
+[package]
+authors = []
+license = "..."
+repository = "..."
+documentation = "..."
+```
+
+These metadata directives are often duplicated amongst all crates, especially
+author/license/repository information. This is a pretty poor experience if you'e
+got to keep writing down the information in so many places!
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Cargo's manifest parsing will be updated with new features to support
+deduplicating each of the areas above. While all of these new features are
+pretty small in their own right, they all add up to greatly reducing the
+overhead of managing a workspace of many crates. The list of new features in
+Cargo will look like the following:
+
+## Workspace-level Dependencies
+
+The `[workspace]` section can now have a `dependencies` section which works the
+same way as the `[dependencies]` section in `Cargo.toml`:
+
+```toml
+# in workspace's Cargo.toml
+[workspace.dependencies]
+log = "0.3.1"
+log2 = { version = "2.0.0", package = "log" }
+serde = { git = 'https://github.com/serde-rs/serde" }
+wasm-bindgen-cli = { path = "crates/cli" }
+```
+
+Each workspace member can then reference this section in the workspace with a
+new dependency directive:
+
+```toml
+# in a workspace member's Cargo.toml
+[dependencies]
+log = { workspace = true }
```

Comment:
How do other keys interact with `workspace = true`? E.g. can I do this?

```toml
log = { workspace = true, features = ["serde"] }
```

Reply:
TBH I didn't really want to consider the interactions of all these keys together, so for now I've got this text:

> The `workspace` key cannot be defined with any other keys other than `optional`.
For example you cannot define both `workspace` and `git`. The value of the
`workspace` key is either `true` or a string. Other values will yield an error.

I'm hoping that this buys us plenty of runway to unlock useful cases now without having to worry too much about directives you've mentioned with `features`. I don't think it would be too too hard to allow other directives, though, but we likely want to consider them on a case-by-case basis.

Some strawman semantics could be that we allow `default-features` and `features` directives, and they "union" with the `[workspace.depenencies]` directive like all feature activations do (`default-features` uses boolean-or and `features` does a set union). 

Do you suspect this'll be a common feature request for `features` specifically to be an allowed key? If so that seems reasonable to me to whitelist as well.

Reply:
Uninoning behavior makes sense IMO.

I think it'll be a reasonably common request to be able to enable features of a shared dependency for one crate or another, but maybe people will just end up configuring the workspace-level dep with the union of all features? That should be realistically fine as long as there's only one "output" of the workspace.

Reply:
Eh I think features have an easy-enough design which should be pretty reasonable here, so I'll add some words which allow them too.